### PR TITLE
implemented astnn network

### PR DESCRIPTION
**Pull request by KatyaKos**
Implemented astnn model. Also added parent class for astnn and tbcnn encoders, and new ast-tokens encoder.

### CODE REVIEW

File: run_trees.sh [hunk 1]
```diff
@@ -18,6 +18,11 @@ python src/train.py $DRYRUN --model tbcnnast \
     resources/data/python/final/jsonl/valid \
     resources/data/python/final/jsonl/test
 
+python src/train.py $DRYRUN --model astnn --sequential trained_models \
```
Comment (niksaz) [inline run_trees.sh:21]: Is the --sequential flag necessary here?

**Review by niksaz — APPROVED** [review]
Thanks, Katya! LGTM.

Could you also add a reference for the wandb.ai model to the README.md with the rest of the models?

File: run_trees.sh (hunk 1 above)
Comment (KatyaKos) [reply run_trees.sh:21]: I'm afraid it is, code and ast encoders both read from the same raw_trees file. You'll get a deadlock without this flag

Comment (niksaz) [reply run_trees.sh:21]: I think you could read simultaneously from the same file in Unix, but let's leave it it like that.

**Comment by KatyaKos** [comment]
Updated Readme with the link to the pre-trained model and some additional detail on the benchmark and our models.

**Pull request merged**